src/analyses/ai_history.h

```diff
@@ -0,0 +1,229 @@
+  ///
+  ///  2. Create a new history object in others and return a pointer to it.
+  ///     Adds the history to the worklist if visit_edge increases the domain.
+  ///     POSTCONDITION(IMPLIES(result.first == step_statust::NEW,
+  ///                           result.second is a new element of others));
+  ///
+  ///  3. Merge with an existing history.
+  ///     Will requeue the history if visit_edge increases the domain.
+  ///     POSTCONDITION(IMPLIES(result.first == step_statust::MERGED,
+  ///                           result.second is an element of others and
+  ///                           others is unchanged));
+  ///
+  ///  4. Block this flow of execution (this may omit some program traces).
+  ///     The abstract interpreter will skip this edge and not queue.
+  ///     POSTCONDITION(IMPLIES(result.first == step_statust::BLOCKED,
+  ///                           result.second == nullptr()));
+
+  /// The order for history_sett
+  virtual bool operator<(const ai_history_baset &op) const = 0;
+
+  /// History objects should be comparable
+  virtual bool operator==(const ai_history_baset &op) const = 0;
+
+  /// The most recent location in the history
+  /// POSTCONDITION(return value is dereferenceable)
+  virtual const locationt &current_location(void) const = 0;
+
+  /// Domains with a substantial height (such as intervals) may need to widen
+  /// when merging this method allows the history to provide a hint on when to
+  /// do this.
+  virtual bool widen(const ai_history_baset &other) const
+  {
+    return false;
+  }
+
+  virtual void output(std::ostream &out) const = 0;
+  virtual jsont output_json(void) const;
+  virtual xmlt output_xml(void) const;
+};
+
+/// The common case of history is to only care about where you are now,
+/// not how you got there!
+/// Invariants are not checkable due to C++...
+class ahistoricalt : public ai_history_baset
+{
+private:
+  // DATA_INVARIANT(current.is_dereferenceable(), "Must not be _::end()")
+  locationt current;
+
+public:
+  ahistoricalt(locationt i) : ai_history_baset(i), current(i)
+  {
+  }
+
+  ahistoricalt(const ahistoricalt &old)
+    : ai_history_baset(old), current(old.current)
+  {
+  }
+
+  /// This will visit all reachable locations at least once.
+  /// Locations will be visited more than once only when this is needed
+  /// for convergence of the domains.
+  step_returnt step(locationt to, history_sett &others) const override
+  {
+    history_ptrt next(new ahistoricalt(to));
+
+    if(others.empty())
+    {
+      const auto r = others.insert(next);
+      CHECK_RETURN(r.second);
+
+      return std::make_pair(step_statust::NEW_FORCE_CONTINUE, next);
+    }
+    else
+    {
+      // Aggressively merge histories because they are indistinguishable
+      INVARIANT(others.size() == 1, "Only needs one history per location");
+
+      const auto it = others.begin();
+      INVARIANT(
+        *(it) == next,
```

Comment:
We're allocating a new `ahistoricalt` just to check this invariant? Maybe only `new` on the others.empty() path and allocate on the stack here?